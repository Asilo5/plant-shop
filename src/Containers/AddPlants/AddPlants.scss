.add-plant-section {
    background: white;
    width: 65%;
    margin: 5em auto;
    height: 75vh;
    box-shadow: 10px 10px 10px #b4b1b1;
    h2 {
        text-align: center;
        padding: 20px 0px 0px;
        font-size: 3em;
    }
    .added-msg {
        font-size: 20px;
        color: green;
        text-decoration: none;
        margin: auto;
        display: block;
        &:hover {
            border-bottom: 1px solid green;
        }
    }
    .error-msg {
      @extend .added-msg;
      color: red;
    }
    form {
        display: flex;
        flex-direction: column;
        padding: 20px;
        width: 60%;
        margin: auto;
        label {
           font-size: 25px;
        }
        input {
           font-size: 20px;
           margin-bottom: 20px;
        }
        button {
            font-size: 28px;
            margin: 20px;
            border-radius: 10px;
            padding: 5px;
            border: 1px solid cadetblue;
            font-weight: bold;
            color: cadetblue;
            &:hover {
                background-color: cadetblue;
                color: white;
                cursor: pointer;
            }
        }
    }
}