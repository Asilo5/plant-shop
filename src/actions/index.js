export const addPlants = (setPlants) => ({
    type: 'ADD_PLANTS',
    setPlants
});

export const hasError = (error) => ({
  type: 'HAS_ERROR',
  error
});

export const newPlant = (plant) => ({
  type: 'NEW_PLANT',
  plant
});

export const successMsg = (msg) => ({
   type: 'SUCCESS_MSG',
   msg
});

export const deletePlant = (id) => ({
  type: 'DELETE_PLANT',
  id
})