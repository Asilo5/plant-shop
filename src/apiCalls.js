export const getPlants = async () => {

      const resp = await fetch('http://localhost:3001');
      const data = await resp.json();

      if(!resp.ok) {
        throw new Error('No Plants Found!');
      } else {
        return data;
      }
   
}