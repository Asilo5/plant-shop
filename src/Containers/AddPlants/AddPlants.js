import React, { Component } from 'react';
import './AddPlants.scss';
import { postPlant } from '../../apiCalls';
import { newPlant, hasError, successMsg } from '../../actions';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';


export class AddPlants extends Component { 
    constructor() {
        super();
        this.state = {
            name: '',
            description: '',
            care: '',
            price: 0,
            image: ''
        }
    }

    handleChange = (e) => {
        this.setState({ [e.target.name] : e.target.value })
        this.setState({ price : parseInt(e.target.value) })
    }

    handleSubmit = (e) => {
      e.preventDefault();

      const {newPlant, hasError, successMsg} = this.props;

       postPlant(this.state)
         .then(data => {
            newPlant(data.plants);
            successMsg('YOUR PLANT IS PUBLISHED, YAY!');
         })
         .catch(err => hasError('Could not submit plant, try again :( '));

       this.clearInputs();
    }

    clearInputs = () => {
        this.setState( {
            name: '',
            description: '',
            care: '',
            price: 0,
            image: ''
        })
    }

    render() {
        return (
            <section className='add-plant-section'>
                <h2>Add Your Plants!</h2>
                <form>
                    <label>Name of Plant:</label>
                    <input type='text'
                           name='name'
                           value={this.state.name}
                           onChange={(e) => this.handleChange(e)}/>
                    <label>Description:</label>
                    <input type='text'
                           name='description'
                           value={this.state.description} 
                           onChange={(e) => this.handleChange(e)}/>
                    <label>Care of Plant:</label>
                    <input type='text'
                           name='care'
                           value={this.state.care} 
                           onChange={(e) => this.handleChange(e)}/>
                    <label>Add Image:</label>
                    <input type='text'
                           name='image'
                           value={this.state.image} 
                           onChange={(e) => this.handleChange(e)}/>
                    <label>Price:</label>
                    <input type='number'
                           name='price'
                           value={this.state.price} 
                           onChange={(e) => this.handleChange(e)}/>
                    <button onClick={(e) => this.handleSubmit(e)}>SUBMIT</button>
                </form>
            </section>
        )
    }
}

export const mapStateToProps = ({succeededMsg}) => ({
    succeededMsg
});

export const mapDispatchToProps = (dispatch) => (
    bindActionCreators({
      newPlant: plant => dispatch(newPlant(plant)),
      hasError: error => dispatch(hasError(error)),
      successMsg: msg => dispatch(successMsg(msg))
    }, dispatch)
);

export default connect(mapStateToProps, mapDispatchToProps)(AddPlants);