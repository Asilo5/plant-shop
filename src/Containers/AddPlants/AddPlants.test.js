import React from 'react';
import { shallow } from 'enzyme';
import { newPlant, hasError, successMsg } from '../../actions';
import { postPlant } from '../../apiCalls';
import { AddPlants, mapStateToProps, mapDispatchToProps } from './AddPlants';

jest.mock('../../apiCalls');

describe('AddPlants container', () => {
  
    describe('AddPlants', () => {
      let wrapper;
      let mockSucceededMsg = 'Posted Plant';
      let mockErrorMsg = 'Error';
      let mockPlant = {
            name: 'Aloe',
            description: 'Green Queen',
            care: 'Loads of love',
            price: 12,
            image: 'image link'
       };

      beforeEach(() => {
          wrapper = shallow(<AddPlants succeededMsg={mockSucceededMsg}
                                       errorMsg={mockErrorMsg} 
                                       newPlant={jest.fn()} 
                                       hasError={jest.fn()} 
                                       successMsg={jest.fn()} />)

         postPlant.mockImplementation(() => {
             return Promise.resolve(mockPlant);
         });
      });

      it('should match snapshot', () => {
          expect(wrapper).toMatchSnapshot();
      })

      it('should call handleChange if input changes on name input', () => {
          wrapper.instance().handleChange = jest.fn();
          const mockEvent = {
              target: {
                name: 'name',
                value: 'Aloe'
              }
          };

          wrapper.find('input').at(0).simulate('change', mockEvent);
          expect(wrapper.instance().handleChange).toHaveBeenCalled();
      })

      it('should call handleChange if input changes on description input', () => {
        wrapper.instance().handleChange = jest.fn();
        const mockEvent = {
            target: {
              name: 'description',
              value: 'Green Queen'
            }
        };

        wrapper.find('input').at(1).simulate('change', mockEvent);
        expect(wrapper.instance().handleChange).toHaveBeenCalled();
      })

      it('should call handleChange if input changes on care input', () => {
        wrapper.instance().handleChange = jest.fn();
        const mockEvent = {
            target: {
              name: 'care',
              value: 'Lots of love'
            }
        };

        wrapper.find('input').at(2).simulate('change', mockEvent);
        expect(wrapper.instance().handleChange).toHaveBeenCalled();
      })

      it('should call handleChange if input changes on image input', () => {
        wrapper.instance().handleChange = jest.fn();
        const mockEvent = {
            target: {
              name: 'image',
              value: 'image link'
            }
        };

        wrapper.find('input').at(3).simulate('change', mockEvent);
        expect(wrapper.instance().handleChange).toHaveBeenCalled();
      })

      it('should call handleChange if input changes on price input', () => {
        wrapper.instance().handleChange = jest.fn();
        const mockEvent = {
            target: {
              name: 'price',
              value: 10
            }
        };

        wrapper.find('input').at(4).simulate('change', mockEvent);
        expect(wrapper.instance().handleChange).toHaveBeenCalled();
      })


      it('should call postPlant when handleSubmit is called', async () => {
          await wrapper.instance().handleSubmit({ preventDefault : jest.fn() });
          expect(postPlant).toHaveBeenCalled(); 

      })

      it('should call on clearInputs when handleSubmit is called', () => {
        wrapper.instance().clearInputs = jest.fn();
        wrapper.instance().handleSubmit({ preventDefault: jest.fn() });
        expect(wrapper.instance().clearInputs).toHaveBeenCalled();
      })

      it('should reset state when clearInput is called', () => {
         const expected = {
            name: '',
            description: '',
            care: '',
            price: 0,
            image: ''
         };

         const filledState = {
            name: 'Aloe',
            description: 'Green queen',
            care: 'loads of love',
            price: 10,
            image: 'image link'
         };

        wrapper.instance().setState(filledState);
        wrapper.instance().clearInputs();
        expect(wrapper.state()).toEqual(expected);
      })

    });

    describe('mapStateToProps', () => {
        
    });

    describe('mapDispatchToProps', () => {
        
    });
})