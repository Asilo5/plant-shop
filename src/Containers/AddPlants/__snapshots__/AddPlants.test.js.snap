// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddPlants container AddPlants should match snapshot 1`] = `
<section
  className="add-plant-section"
>
  <h2>
    Add Your Plants!
  </h2>
  <NavLink
    className="added-msg"
    to="/allPlants"
  >
    Posted Plant
  </NavLink>
  <form>
    <label>
      Name of Plant:
    </label>
    <input
      name="name"
      onChange={[Function]}
      type="text"
      value=""
    />
    <label>
      Description:
    </label>
    <input
      name="description"
      onChange={[Function]}
      type="text"
      value=""
    />
    <label>
      Care of Plant:
    </label>
    <input
      name="care"
      onChange={[Function]}
      type="text"
      value=""
    />
    <label>
      Add Image:
    </label>
    <input
      name="image"
      onChange={[Function]}
      type="text"
      value=""
    />
    <label>
      Price:
    </label>
    <input
      name="price"
      onChange={[Function]}
      type="number"
      value={0}
    />
    <button
      onClick={[Function]}
    >
      SUBMIT
    </button>
  </form>
</section>
`;
